// @salla.sa/twilight-bundles/dist/twilight-bundles.js
import { LitElement as g, css as f } from "lit";
import { property as l } from "lit/decorators.js";
class r {
  static onBundlesReady() {
    return r.makeSureSallaIsReady().then(() => Salla.event.onlyWhen("twilight-bundles::initiated")).then(() => Salla.lang.onLoaded());
  }
  static async initializeSalla() {
    if (Salla.status === "ready") {
      salla.log("Salla is ready");
      return;
    }
    const e = document.currentScript || document.querySelector('script[src*="twilight-bundles.js"]'), t = e == null ? void 0 : e.hasAttribute("demo-mode");
    let n = e == null ? void 0 : e.getAttribute("store-id"), i = JSON.parse((e == null ? void 0 : e.getAttribute("config")) || "false");
    return t || i || n ? (n = n || "1510890315", i = i || await r.getStoreSettings(n), Salla.init(i || {
      debug: !0,
      store: { id: n }
    })) : Salla.onReady();
  }
  static async getStoreSettings(e) {
    let t = Salla.storage.session.get(`store-settings-${e}`);
    return t || (t = (await fetch("https://api.salla.dev/store/v1/store/settings", { headers: { "store-identifier": e } }).then((n) => n.json())).data, Salla.storage.session.set(`store-settings-${e}`, t), t);
  }
  static makeSureSallaIsReady() {
    return window.Salla ? Promise.resolve(r.initializeSalla()) : new Promise((e, t) => {
      let n;
      const i = setTimeout(() => {
        window.clearInterval(n), t(new Error("Timeout: Salla object not found after 10 seconds"));
      }, 1e4);
      n = window.setInterval(() => {
        window.Salla && (window.clearInterval(n), clearTimeout(i), e(r.initializeSalla()));
      }, 50);
    });
  }
}
var y = Object.defineProperty, d = (m, e, t, n) => {
  for (var i = void 0, s = m.length - 1, o; s >= 0; s--)
    (o = m[s]) && (i = o(e, t, i) || i);
  return i && y(e, t, i), i;
};
const c = class c extends g {
  constructor() {
    super(...arguments), this.key = "", this.data = {}, this.imports = {}, this.shadowRootMode = "open";
  }
  createRenderRoot() {
    return this.shadowRootMode !== !1 ? this.attachShadow({ mode: this.shadowRootMode || "closed" }) : this;
  }
  static register(e) {
    const t = {
      component: this,
      dynamicTagName: `${e}-${Math.random().toString(36).substring(2, 8)}`
    };
    return r.onBundlesReady().then(() => Salla.bundles.registerComponent(e, t));
  }
};
c.styles = f`:host { display: block; }`;
let a = c;
d([
  l()
], a.prototype, "key");
d([
  l()
], a.prototype, "data");
d([
  l()
], a.prototype, "imports");
d([
  l()
], a.prototype, "shadowRootMode");
const h = "twilight-skeleton-styles", S = ["min-height", "height", "width", "aspect-ratio"];
class u extends HTMLElement {
  connectedCallback() {
    var t;
    let e = (t = this.getAttribute("component-name")) == null ? void 0 : t.replace(/^salla-/, "");
    if (!e)
      return Salla.error("Component name is required", this), Promise.resolve();
    this.renderSkeleton(e), Salla.bundles.renderCustomComponentDom(`salla-${e}`, this);
  }
  renderSkeleton(e) {
    var i;
    if (this.dataset.skeletonRendered === "true" || (this.dataset.skeletonRendered = "true", this.hasMeaningfulChildren()) || (i = Salla.bundles) != null && i.isRegistered(`salla-${e}`)) return;
    u.ensureSkeletonStyles();
    const t = document.createElement("div");
    t.className = "twilight-skeleton", t.setAttribute("part", "skeleton"), t.setAttribute("aria-hidden", "true");
    let n = !1;
    S.forEach((s) => {
      const o = this.getAttribute(s);
      o && (t.style.setProperty(s, o), n = !0);
    }), n || (t.style.minHeight = "1.5em"), this.replaceChildren(t);
  }
  hasMeaningfulChildren() {
    for (const e of Array.from(this.childNodes))
      if (e.nodeType === Node.ELEMENT_NODE || e.nodeType === Node.TEXT_NODE && (e.nodeValue || "").trim().length > 0) return !0;
    return !1;
  }
  static ensureSkeletonStyles() {
    if (document.getElementById(h)) return;
    const e = document.createElement("style");
    e.id = h, e.textContent = `
.twilight-skeleton {
  display: block;
  width: 100%;
  background: linear-gradient(90deg, #eeeeee 0%, #f5f5f5 50%, #eeeeee 100%);
  background-size: 200% 100%;
  animation: twilight-skeleton-shimmer 1.4s ease-in-out infinite;
  border-radius: 4px;
}
@keyframes twilight-skeleton-shimmer {
  0%   { background-position: 200% 0; }
  100% { background-position: -200% 0; }
}
@media (prefers-reduced-motion: reduce) {
  .twilight-skeleton { animation: none; }
}
`, document.head.appendChild(e);
  }
}
class w {
  constructor() {
    this.components = /* @__PURE__ */ new Map(), this.pendingComponents = [], this.initialized = !1, Salla.onReady().then(() => {
      if (Salla.bundles) {
        Salla.log("TwilightBundles is already initialized");
        return;
      }
      this.init(), Salla.bundles = this, Salla.event.emit("twilight-bundles::initiated"), this.registerCustomComponents();
    });
  }
  async init() {
    this.initialized || (this.initialized = !0);
  }
  registerCustomComponents() {
    HTMLElement.registerSallaComponent = function(e) {
      Salla.bundles.registerComponent(e, {
        component: this,
        dynamicTagName: `${e}-${Math.random().toString(36).substring(2, 8)}`
      });
    }, this.injectBundleScript(window.customComponents || []);
  }
  injectBundleScript(e) {
    e == null || e.forEach((t) => {
      if (document.querySelector(`script[src="${t}"]`)) {
        Salla.log("Script already exists", t);
        return;
      }
      const n = document.createElement("script");
      n.type = "module", n.src = t, document.head.appendChild(n);
    });
  }
  /**
   * This is takes the `<salla-custom-component component-name="${tagName}">` element
   * and replaces it with the actual component
   */
  renderCustomComponentDom(e, t) {
    Salla.log("Rendering custom component", e);
    const n = this.components.get(e);
    if (n)
      return this.renderDynamicCustomComponentDom(n.dynamicTagName, t);
    this.pendingComponents.push({ tagName: e, component: t });
    const i = t.getAttribute("version"), s = t.getAttribute("assets");
    if (s) {
      this.injectBundleScript(s.split(",").map((p) => this.appendVersion(p, i)));
      return;
    }
    const o = t.getAttribute("bundle-id");
    if (o) {
      this.injectBundleScript([this.appendVersion(this.getAssetForComponent(o, e.replace(/^salla-/, "")), i)]);
      return;
    }
  }
  appendVersion(e, t) {
    if (!t) return e;
    const n = e.includes("?") ? "&" : "?";
    return `${e}${n}ver=${t}`;
  }
  getAssetForComponent(e, t) {
    var n, i;
    return (i = (n = Salla.config.get("theme.bundle_assets", "https://cdn.assets.salla.network/themes/:bundle/latest/:component.js")) == null ? void 0 : n.replace(":bundle", e)) == null ? void 0 : i.replace(":component", t);
  }
  renderDynamicCustomComponentDom(e, t) {
    const n = document.createElement(e);
    Array.from(t.attributes).forEach((i) => n.setAttribute(i.name, i.value || "")), ["min-height", "height", "width", "aspect-ratio"].forEach((i) => {
      const s = t.getAttribute(i);
      s && n.style.setProperty(i, s);
    }), t.before(n), t.remove(), requestAnimationFrame(() => requestAnimationFrame(() => {
      n.style.removeProperty("min-height"), n.style.removeProperty("aspect-ratio"), n.setAttribute("loaded", "");
    }));
  }
  isRegistered(e) {
    return this.components.has(e);
  }
  registerComponent(e, t) {
    if (this.components.has(e))
      return console.warn(`Component ${e} is already registered into the this.components map. Skipping.`);
    if (!t.dynamicTagName)
      return console.warn(`Component ${e} is missing dynamicTagName. Skipping.`);
    if (!t.component)
      return console.warn(`Component ${e} is missing component. Skipping.`);
    if (window.customElements.get(t.dynamicTagName))
      return console.warn(`Component ${t.dynamicTagName} is already registered into the window custom elements. Skipping.`);
    window.customElements.define(t.dynamicTagName, t.component), this.components.set(e, t), Salla.log("Component registered:", t.dynamicTagName);
    const n = this.pendingComponents.filter((i) => i.tagName === e);
    n.length && (n.forEach((i) => {
      this.renderDynamicCustomComponentDom(t.dynamicTagName, i.component);
    }), this.pendingComponents = this.pendingComponents.filter((i) => i.tagName !== e), Salla.log("Pending Components rendered:", e, t.dynamicTagName));
  }
  renderSharedAssests(...e) {
    const t = {
      sallaicons: `<link rel="stylesheet" href="${Salla.url.cdn("fonts/sallaicons.css")}">`
    };
    return e.map((n) => t[n] || "").join("");
  }
}
r.makeSureSallaIsReady().then(() => new w()).then(() => window.customElements.get("salla-custom-component") || window.customElements.define("salla-custom-component", u));
export {
  a as SallaComponent
};
